daemon/containerd/image_push.go

```diff
@@ -2,15 +2,277 @@ package containerd
+	return i.pushImage(ctx, targetRef, platform, metaHeaders, authConfig, outStream)
+}
+
+// PushImage initiates a push operation of the image pointed to by targetRef.
+//
+// If image points to an index/manifest list and platform is specified, then only
+// a matching platform manifest is pushed.
+// Otherwise index is pushed as is, which will probably fail if you don't have all
+// content referenced by the index. Cross-repo mounts will be attempted for non-existing blobs.
+func (i *ImageService) pushImage(ctx context.Context, targetRef reference.Named, optPlatform *ocispec.Platform, metaHeaders map[string][]string, authConfig *registry.AuthConfig, outStream io.Writer) error {
+	if _, ok := targetRef.(reference.Tagged); !ok {
+		return errdefs.NotImplemented(errors.New("push all tags is not implemented"))
+	}
+
+	leasedCtx, release, err := i.client.WithLease(ctx)
+	if err != nil {
+		return err
+	}
+	defer release(leasedCtx)
+	out := streamformatter.NewJSONProgressOutput(outStream, false)
+
+	img, err := i.client.ImageService().Get(ctx, targetRef.String())
+	if err != nil {
+		return errdefs.NotFound(err)
+	}
+
+	target := img.Target
+	store := i.client.ContentStore()
+
+	if optPlatform != nil {
+		platform := *optPlatform
+		pm := platforms.OnlyStrict(platform)
+		if img.Target.Platform != nil && !pm.Match(*img.Target.Platform) {
+			return errdefs.NotFound(fmt.Errorf(
+				"image platform (%s) doesn't match the requested platform (%s)",
+				platforms.Format(*img.Target.Platform), platforms.Format(platform)))
+		}
+
+		if containerdimages.IsIndexType(img.Target.MediaType) {
+			getManifests := containerdimages.LimitManifests(
+				containerdimages.FilterPlatforms(containerdimages.ChildrenHandler(store), pm), pm, 0)
+
+			manifests, err := getManifests(ctx, img.Target)
+			if err != nil {
+				if cerrdefs.IsNotFound(err) {
+					return errdefs.NotFound(err)
+				}
+				return errdefs.Unknown(errors.Wrapf(err, "failed to get manifests"))
+			}
+
+			presentManifests := make([]ocispec.Descriptor, 0, 1)
+			for _, mnfst := range manifests {
+				_, err := store.Info(ctx, mnfst.Digest)
+				if err == nil {
+					presentManifests = append(presentManifests, mnfst)
+				}
+			}
+
+			switch len(presentManifests) {
+			case 1:
+				target = presentManifests[0]
+			case 0:
+				return errdefs.NotFound(fmt.Errorf("no matching image manifest found for platform %s", platforms.Format(platform)))
+			default:
+				return errdefs.InvalidParameter(fmt.Errorf("more than one manifest matches the desired platform %s", platforms.Format(platform)))
+			}
+		}
+	}
+
+	resolver, tracker := i.newResolverFromAuthConfig(authConfig)
+	pushProgress := pushProgress{Tracker: tracker}
+	jobs := newJobs()
+	finishProgress := jobs.showProgress(ctx, out, combinedProgress([]progressUpdater{
+		&pushProgress,
+		pullProgress{ShowExists: false, Store: store},
+	}))
+	defer finishProgress()
+
+	var limiter *semaphore.Weighted = nil // TODO: Respect max concurrent downloads/uploads
```

Comment:
Actually we can support it for push just fine because we use the `remotes.PushContent` directly. It's the pull that uses the c8d's `client.Push` method that can't use the same limiter for multiple concurrent operations. 
For pull, I expect that at some point we would have the need to do something more than the c8d `client.Pull` provides and we will have to use the lower level functions anyway, which would allow us to support MaxConcurrentDownloads too.
So perhaps it would be fine (at least for now) to only have "MaxConcurrentUploads" implemented?

Reply:
Yeah, that seems reasonable to me, although IMO doesn't have to be part of _this_ PR unless you think the implementation is really straightforward/quick. :bow:

Given this isn't blocked on upstream (as I'd mistakenly crossed the upload/download signals in my head), probably fine to leave the comment as-is too. :+1: